### DISCUSSION TITLE
Closest language to Project Verona?

### DISCUSSION THREAD

**Discussion by WashingtonARamos** [microsoft/verona#416] [Q&A] [answered]
Hi, which language could be considered the closest to what Project Verona aims to be at "release"?

Will it run on .NET (like a C# focused on concurrency), compile to native but have a garbage collection mechanism (like Go) or be a systems programming language to compete with Rust, C and the likes?

**Comment by davidchisnall** [comment, marked as answer]
Hi,

This question is a bit difficult to answer.  Verona is a research project and it exists because we believe that there's a large space in language design that is not explored by existing languages.

Verona is not going to run on .NET or any other VM, it will be natively compiled.  It will not have a global garbage collector, each region can have a separate memory management policy (which may be a tracing GC, reference counting, bump-allocation and bulk free, or other strategies).  The type system statically provides the non-interference guarantees that are necessary for this to work and ensure that no code that does not make use of traced regions is impacted by the tracing.  

Verona is an infrastructure language, not a systems language.  Unlike Rust and C, it does not provide any mechanisms that allow violating type safety, even for interop with unsafe languages.  These escape hatches are necessary in a language such as C/C++ that is used to implement context-switch routines, memory managers, language runtimes, and other things of this nature that explicitly require violating type- and memory-safety but come with a large cost: You must either be pessimistic in optimisations because some code can break your language's rules, or you must accept 'undefined behaviour', meaning that some thing that a programmer can do are able to completely break unrelated parts of the program in unexpected ways.

Libraries in unsafe languages can be imported and exposed directly to Verona, but their heap is exposed as a single region allowing various mechanisms (SFI, CHERI, process / MMU-based isolation) to ensure that memory safety bugs in C/C++/assembly code cannot impact the type- and concurrency-safety guarantees that Verona provides.
